Added example and documentation for fips compliant communication with Kafka cluster (#1582)
confluentinc/confluent-kafka-python: examples/fips/README.md

# FIPS Compliance

## Communication between client and Kafka cluster

### Install from prebuilt wheels

> If this client is installed using `pip`, it comes with librdkafka in which OpenSSL 3.0 is statically linked. This client can communicates with Kafka cluster with FIPS approved algorithms using FIPS provider provided by OpenSSL.

Reviewer: Here, is the link to "client-configuration-to-enable-fips-provider" required? We need to still change the client configuration, right?

Or do we need to skip to another section instead?

Author: The `enabling FIPS provider` section is needed for this. I think I will make it more clear.

### Install from source by building librdkafka

When building librdkafka and python client from source, librdkafka uses OpenSSL present in the system by dynamic linking it. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to section [Client configuration to enable fips provider](#client-configuration-to-enable-fips-provider), otherwise use below steps to make system OpenSSL FIPS compliance. Once that is done, librdkafka and python client will use FIPS approved algorithms for the communication between client and Kafka Cluster using producer, consumer or admin client.

### Enabling FIPS provider

> We assume that you have FIPS provider (module) available in your system. This module can be dynamically plugged into OpenSSL by putting this module in default module folder of OpenSSL or pointing this module with environment vairable `OPENSSL_MODULES` (folder containing the fips module in the form of dynamic library). We also need to use OpenSSL config which has FIPS configuration present in it. Default configuration file can be modified to include FIPS related config or a new configuration file can also be pointed using `OPENSSL_CONF` environment variable.

Reviewer: > pointing this module

pointing to this module

> We assume that you have FIPS provider (module) available in your system. This module can be dynamically plugged into OpenSSL by putting this module in default module folder of OpenSSL or pointing this module with environment vairable `OPENSSL_MODULES` (folder containing the fips module in the form of dynamic library). We also need to use OpenSSL config which has FIPS configuration present in it. Default configuration file can be modified to include FIPS related config or a new configuration file can also be pointed using `OPENSSL_CONF` environment variable.

Author: Good point. We are testing it with OpenSSL 3.0 and hence we are providing all the instructions related to it. But I think we should point out that this library is FIPS compliant with OpenSSL 3.0 and lesser version should be working in similar way but we can't guarantee anything as we have not tested with older versions.

#### Steps to generate FIPS provider

Steps to generate fips provider can be found on the [README-FIPS doc](https://github.com/openssl/openssl/blob/openssl-3.0.8/README-FIPS.md)

In short, you need to perform the following steps:

1) Clone OpenSSL from [OpenSSL Github Repo](https://github.com/openssl/openssl)
2) Checkout the correct version. (v3.0.8 is the current fips compliant version for OpenSSL 3.0 at the time of writing this doc.)
3) Run `./configure --enable-fips`
4) Run `make install_fips`

After last step, two files are generated.
* fips module (`fips.dylib` in Mac, `fips.so` in Linux and `fips.dll` in Windows)
* fips config (`fipsmodule.cnf`) file will be generated. Which needs to be used with OpenSSL.

#### Referencing FIPS provider in OpenSSL

As mentioned earliar, the above generated fips module can be dynamically plugged into OpenSSL by putting this module in default module folder (Refer [this](https://github.com/confluentinc/librdkafka/blob/master/INTRODUCTION.md#ssl) for typical default locations on various OS) of OpenSSL. It should look something like `...lib/ossl-modules/`.

> This module can also be pointed with environment vairable `OPENSSL_MODULES` (folder containing the fips module in the form of dynamic library)

Reviewer: variable

#### Linking FIPS provider with OpenSSL

`fipsmodule.cnf` file includes `fips_sect` which is required to enable fips in OpenSSL. `fipsmodule.cnf` needs to be included in `openssl.cnf` file to make it work.

Some of the algorithms might have different implementation in fips or other providers. If you load two different providers like default and fips, any implementation could be used. To make sure the we fetch only fips compliant version of the algorithm, we use `fips=yes` default property in config file.

OpenSSL config should look something like after applying the above changes.

```
config_diagnostics = 1
openssl_conf = openssl_init

.include /usr/local/ssl/fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
fips = fips_sect

[algorithm_sect]
default_properties = fips=yes
.
.
.
```

### Client configuration to enable fips provider

> Some of the non crypto algorithms are required in OpenSSL. These algorithms are not included in fips provider and hence we need to use `base` provider in conjunction with `fips` provider. Base provider comes with OpenSSL by default. We just need to enable it in the client configuration.

Reviewer: I think it should say "Some of the non-crypto OpenSSL algorithms are required by the client"

Author: It's not specific to OpenSSL for example base64 encoding.

# FIPS Compliance

> We are testing and making this client FIPS compliant using OpenSSL 3.0. Use OpenSSL 3.0 to use this client in FIPS compliant mode. Older versions of OpenSSL might work in similar way but we have not verified that.

Reviewer 2: ```suggestion
We tested FIPS compliance for the client using OpenSSL 3.0. To use the client in FIPS-compliant mode, use OpenSSL 3.0. Older versions of OpenSSL have not been verified (although they may work).
```

Author: Used suggestion

## Communication between client and Kafka cluster

### Install from prebuilt wheels

> If this client is installed using `pip`, it comes with librdkafka in which OpenSSL 3.0 is statically linked. This client can communicates with Kafka cluster with FIPS approved algorithms using FIPS provider provided by OpenSSL. Steps to enable FIPS provider is mentioned in section [Enabling FIPS provider](#enabling-fips-provider).

Reviewer 2: ```suggestion
If you install this client using `pip`, you get the client with librdkafka with a static link to OpenSSL 3.0. To enable this client to communicate with the Kafka cluster using the OpenSSL FIPS provider and FIPS-approved algorithms, you must enable the FIPS provider. You can find steps to enable the FIPS provider in section [Enabling FIPS provider](#enabling-fips-provider).
```
I'm guessing that the user installs and then needs to enable? If not, let me know and I can rework this section.

Author: Used your suggestion with some changes. Please check.

### Install from source by building librdkafka

> When building librdkafka and python client from source, librdkafka uses OpenSSL present in the system by dynamically linking it. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to section [Client configuration to enable FIPS provider](#client-configuration-to-enable-fips-provider), otherwise use steps mentioned in section [Enabling FIPS provider](#enabling-fips-provider) to make system OpenSSL FIPS compliant. Once that is done, librdkafka and python client will use FIPS approved algorithms for the communication between client and Kafka Cluster using producer, consumer or admin client.

Reviewer 2: ```suggestion
When you build the python client from source, librdkafka dynamically links to the OpenSSL present in the system. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to section [Client configuration to enable FIPS provider](#client-configuration-to-enable-fips-provider), otherwise use steps mentioned in section [Enabling FIPS provider](#enabling-fips-provider) to make OpenSSL in your system FIPS compliant. Once that is done, librdkafka and python client will use FIPS approved algorithms for the communication between client and Kafka Cluster using producer, consumer or admin client.
```

Author: Used your suggestion with some changes. Please check.

Reviewer 3: The change to "When you build the python client from source" is ok because one could choose to install the binary librdkafka linux package and just compile or install OpenSSL. Probably it's the better alternative to just using the wheels.

### Enabling FIPS provider

> We assume that you have FIPS provider (module) available in your system. This module can be dynamically plugged into OpenSSL by putting this module in default module folder of OpenSSL or pointing to this module with environment vairable `OPENSSL_MODULES` (folder containing the FIPS module in the form of dynamic library). We also need to use OpenSSL config which has FIPS configuration present in it. Default configuration file can be modified to include FIPS related config or a new configuration file can also be pointed using `OPENSSL_CONF` environment variable.

Reviewer 2: ```suggestion
The FIPS provider is a module. To enable the FIPS provider, you must have the module available on your system, plug the module into OpenSSL, and then configure OpenSSL to use the module. 

You can plug the FIPS provider into OpenSSL two ways: 1) put the module in the default module folder of OpenSSL or 2) point to the module with the environment variable, `OPENSSL_MODULES`. For example: `OPENSSL_MODULES="/path/to/fips/module/lib/folder/` 

You configure OpenSSL to use the FIPS provider using the FIPS configuration in OpenSSL config. You can modify the default configuration file to include FIPS related config or create a new configuration file point to it using the environment variable,`OPENSSL_CONF`. For example `OPENSSL_CONF="/path/to/fips/enabled/openssl/config/openssl.cnf`: 
```
This might be improved by adding what needs to be configured in OpenSSL config.

Author: Used your suggestion with some changes. Please check.

Reviewer 2: @pranavrth the changes above look good to me. I have looked and have no further recommendations.

#### Steps to generate FIPS provider

> Steps to generate FIPS provider can be found on the [README-FIPS doc](https://github.com/openssl/openssl/blob/openssl-3.0.8/README-FIPS.md)

Reviewer 2: ```suggestion
You can find steps to generate the FIPS provider in the [README-FIPS doc](https://github.com/openssl/openssl/blob/openssl-3.0.8/README-FIPS.md)
```

Author: Used suggestion

In short, you need to perform the following steps:

1) Clone OpenSSL from [OpenSSL Github Repo](https://github.com/openssl/openssl)
2) Checkout the correct version. (v3.0.8 is the current FIPS compliant version for OpenSSL 3.0 at the time of writing this doc.)
3) Run `./configure --enable-fips`
4) Run `make install_fips`

After last step, two files are generated.
* FIPS module (`fips.dylib` in Mac, `fips.so` in Linux and `fips.dll` in Windows)
* FIPS config (`fipsmodule.cnf`) file will be generated. Which needs to be used with OpenSSL.

#### Referencing FIPS provider in OpenSSL

> As mentioned earlier, the above generated FIPS module can be dynamically plugged into OpenSSL by putting this module in default module folder (Refer [this](https://github.com/confluentinc/librdkafka/blob/master/INTRODUCTION.md#ssl) for typical default locations on various OS) of OpenSSL. It should look something like `...lib/ossl-modules/`.

Reviewer 2: ```suggestion
As mentioned earlier, you can dynamically plug the FIPS module generated above into OpenSSL by putting the FIPS module into the default OpenSSL module folder. For default locations of OpenSSL on various operating systems, see the SSL section of the [Introduction to librdkafka - the Apache Kafka C/C++ client library](https://github.com/confluentinc/librdkafka/blob/master/INTRODUCTION.md#ssl). It should look something like `...lib/ossl-modules/`.
```

Author: Used Suggestion.

> This module can also be pointed with environment variable `OPENSSL_MODULES` (folder containing the FIPS module in the form of dynamic library)

Reviewer 2: ```suggestion
You can also point to this module with the environment variable `OPENSSL_MODULES` (folder containing the FIPS module in the form of dynamic library)
```

Author: Used suggestion

#### Linking FIPS provider with OpenSSL

> `fipsmodule.cnf` file includes `fips_sect` which is required to enable FIPS in OpenSSL. `fipsmodule.cnf` needs to be included in `openssl.cnf` file to make it work.

Reviewer 2: ```suggestion
To enable FIPS in OpenSSL, you must include `fipsmodule.cnf` in the file, `openssl.cnf`. The `fipsmodule.cnf` file includes `fips_sect` which OpenSSL requires to enable FIPS. See the example below. 
```
I struggled to understand this part. I am happy to look at it again. The main thing is to make clear what the customer has to do to get it to work.

Author: Used suggestion. It looks fine.

> Some of the algorithms might have different implementation in FIPS or other providers. If you load two different providers like default and fips, any implementation could be used. To make sure the we fetch only FIPS compliant version of the algorithm, we use `fips=yes` default property in config file.

Reviewer 2: ```suggestion
Some of the algorithms might have different implementation in FIPS or other providers. If you load two different providers like default and fips, any implementation could be used. To make sure you fetch only FIPS compliant version of the algorithm, use `fips=yes` default property in config file.
```

Author: Used suggestion

OpenSSL config should look something like after applying the above changes.

```
config_diagnostics = 1
openssl_conf = openssl_init

.include /usr/local/ssl/fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
fips = fips_sect

[algorithm_sect]
default_properties = fips=yes
.
.
.
```

### Client configuration to enable FIPS provider

> Some of the non-crypto algorithms are required in OpenSSL as well. These algorithms are not included in FIPS provider and hence we need to use `base` provider in conjunction with `fips` provider. Base provider comes with OpenSSL by default. We just need to enable it in the client configuration.

Reviewer 2: ```suggestion
OpenSSL requires some non-crypto algorithms as well. These algorithms are not included in the FIPS provider and you need to use the `base` provider in conjunction with the `fips` provider. Base provider comes with OpenSSL by default. You must enable `base` provider in the client configuration.
```

Author: Used suggestion

> To make client (consumer, producer or admin client) FIPS compliant, we need to enable only `fips` and `base` provider in the client using `ssl.providers: 'fips,base'` property.

Reviewer 2: ```suggestion
To make client (consumer, producer or admin client) FIPS compliant, you must enable only `fips` and `base` provider in the client using the `ssl.providers` property. For example: `ssl.providers: 'fips,base'`
```

Author: Used your suggestion with some changes. Please check.

# FIPS Compliance

We tested FIPS compliance for the client using OpenSSL 3.0. To use the client in FIPS-compliant mode, use OpenSSL 3.0. Older versions of OpenSSL have not been verified (although they may work).

## Communication between client and Kafka cluster

### Install from prebuilt wheels

> If you install the this client through prebuilt wheels using `pip`, it comes with librdkafka in which OpenSSL 3.0 is already statically linked. To enable this client to communicate with the Kafka cluster using the OpenSSL FIPS provider and FIPS-approved algorithms, you must enable the FIPS provider. You can find steps to enable the FIPS provider in section [Enabling FIPS provider](#enabling-fips-provider).

Reviewer 3: "the this" should be just "the" or "this"

> If you install the this client through prebuilt wheels using `pip`, it comes with librdkafka in which OpenSSL 3.0 is already statically linked. To enable this client to communicate with the Kafka cluster using the OpenSSL FIPS provider and FIPS-approved algorithms, you must enable the FIPS provider. You can find steps to enable the FIPS provider in section [Enabling FIPS provider](#enabling-fips-provider).

Reviewer 2: ```suggestion
If you install the client through prebuilt wheels using `pip`, it comes with librdkafka and OpenSSL 3.0 is already statically linked. To enable this client to communicate with the Kafka cluster using the OpenSSL FIPS provider and FIPS-approved algorithms, you must enable the FIPS provider. You can find steps to enable the FIPS provider in section [Enabling FIPS provider](#enabling-fips-provider).
```

Author: This is changed alot and I think will change more. Let me consider this after all the changes.

> ### Install from source by building librdkafka

Author: I am planning to change both the headings. From 
1. _`Install from prebuilt wheels`_ to **`Use statically linked OpenSSL with librdkafka shared library (prebuilt wheels or prebuilt librdkafka library)`**
2.  _`Install from source by building librdkafka`_  to **`Use installed OpenSSL version with librdkafka shared library (installing librdkafka from source)`**

> ### Install from source by building librdkafka

Reviewer 3: I'd say
1. "Use statically linked OpenSSL with librdkafka bundled in wheels"

> ### Install from source by building librdkafka

Author: I think we only package statically linked OpenSSL for all linux distributions. I was checking packing build in Semaphore.

> ### Install from source by building librdkafka

Reviewer 3: So, we cannot suggest to use those packages as they're linked with OSSL 1, we need to distribute a librdkafka2 that links to OSSL 3. Let's suggest to compile librdkafka too, if not using wheels

> When you build the python client from source, librdkafka dynamically links to the OpenSSL present in the system. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to section [Client configuration to enable FIPS provider](#client-configuration-to-enable-fips-provider) to enable `fips` provider, otherwise use steps mentioned in section [Enabling FIPS provider](#enabling-fips-provider) to make OpenSSL in your system FIPS compliant and then enable `fips` provider. Once that is done, librdkafka and python client will use FIPS approved algorithms for the communication between client and Kafka Cluster using producer, consumer or admin client.

Reviewer 3: Add that if you want to use system installed librdkafka and OpenSSL it's necessary to install confluent_kafka with:
```
pip install confluent_kafka --no-binary :all:
```

> When you build the python client from source, librdkafka dynamically links to the OpenSSL present in the system. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to section [Client configuration to enable FIPS provider](#client-configuration-to-enable-fips-provider) to enable `fips` provider, otherwise use steps mentioned in section [Enabling FIPS provider](#enabling-fips-provider) to make OpenSSL in your system FIPS compliant and then enable `fips` provider. Once that is done, librdkafka and python client will use FIPS approved algorithms for the communication between client and Kafka Cluster using producer, consumer or admin client.

Reviewer 2: ```suggestion
When you build the python client from source, librdkafka dynamically links to the OpenSSL present in the system. If the installed OpenSSL is already working in FIPS mode, then you can directly jump to the section [Client configuration to enable FIPS provider](#client-configuration-to-enable-fips-provider) and enable the `fips` provider. If you don't have OpenSSL working in FIPS mode, use the steps mentioned in the section [Enabling FIPS provider](#enabling-fips-provider) to make OpenSSL in your system FIPS compliant and then enable the `fips` provider. Once OpenSSL working in FIPS mode and the `fips` provider is enabled, librdkafka and python client will use the FIPS approved algorithms for communication between client and Kafka Cluster using the producer, consumer or admin client.
```

### Enabling FIPS provider

To enable the FIPS provider, you must have the FIPS module available on your system, plug the module into OpenSSL, and then configure OpenSSL to use the module. 

> You can plug the FIPS provider into OpenSSL two ways: 1) put the module in the default module folder of OpenSSL or 2) point to the module with the environment variable, `OPENSSL_MODULES`. For example: `OPENSSL_MODULES="/path/to/fips/module/lib/folder/` 

Reviewer 3: You could specify that 1 can be used only when installing from source, as with the prebuild wheel it will search a path that corresponds to the Semaphore OpenSSL source folder. With the prebuild wheel only option 2 can be used, if I'm correct.

> You configure OpenSSL to use the FIPS provider using the FIPS configuration in OpenSSL config. You can modify the default configuration file to include FIPS related config or create a new configuration file and point to it using the environment variable,`OPENSSL_CONF`. For example `OPENSSL_CONF="/path/to/fips/enabled/openssl/config/openssl.cnf` For an example of OpenSSL config, see below. 

Reviewer 3: Same thing with the configuration file probably

> #### Steps to generate FIPS provider module

Reviewer 3: Could I be better to say to "build FIPS provider module"?

You can find steps to generate the FIPS provider module in the [README-FIPS doc](https://github.com/openssl/openssl/blob/openssl-3.0.8/README-FIPS.md)

In short, you need to perform the following steps:

1) Clone OpenSSL from [OpenSSL Github Repo](https://github.com/openssl/openssl)
2) Checkout the correct version. (v3.0.8 is the current FIPS compliant version for OpenSSL 3.0 at the time of writing this doc.)

> 3) Run `./configure --enable-fips`

Author: I think `--prefix` option is confusing in our case. User will check all the options other than `enable-fips` on net if required.

I am correcting the command.

4) Run `make install_fips`

After last step, two files are generated.
* FIPS module (`fips.dylib` in Mac, `fips.so` in Linux and `fips.dll` in Windows)
* FIPS config (`fipsmodule.cnf`) file will be generated. Which needs to be used with OpenSSL.

#### Referencing FIPS provider in OpenSSL

> As mentioned earlier, you can dynamically plug the FIPS module generated above into OpenSSL by putting the FIPS module into the default OpenSSL module folder. For default locations of OpenSSL on various operating systems, see the SSL section of the [Introduction to librdkafka - the Apache Kafka C/C++ client library](https://github.com/confluentinc/librdkafka/blob/master/INTRODUCTION.md#ssl). It should look something like `...lib/ossl-modules/`.

Reviewer 3: add a "only when installing from source" specification here too
